diff --git a/frontend/test/pytest/test_template.py b/frontend/test/pytest/test_template.py
index ee2ecfeb0..e3a8c6bcf 100644
--- a/frontend/test/pytest/test_template.py
+++ b/frontend/test/pytest/test_template.py
@@ -81,7 +81,7 @@ def test_basis_embedding(backend):
     """Test basis embedding."""
 
     def basis_embedding(f: jax.core.ShapedArray([3], int)):
-        qp.BasisEmbedding(features=f, wires=[0, 1, 2])
+        qp.BasisEmbedding(f, wires=[0, 1, 2])
         return qp.state()
 
     device = qp.device(backend, wires=3)
