--- build/python_packages/quantum/mlir_quantum/dialects/_tensor_ops_gen.py	2023-03-07 13:06:45.209375306 -0500
+++ build/python_packages/quantum/mlir_quantum/dialects/_tensor_ops_gen.py.backup	2023-03-07 13:06:36.653489695 -0500
@@ -266,6 +266,26 @@
 
   _ODS_REGIONS = (0, True)
 
+  def __init__(self, result, values, *, loc=None, ip=None):
+    operands = []
+    results = []
+    attributes = {}
+    regions = None
+    operands.extend(_get_op_results_or_values(values))
+    results.append(result)
+    _ods_successors = None
+    super().__init__(
+        self.build_generic(
+            attributes=attributes,
+            results=results,
+            operands=operands,
+            successors=_ods_successors,
+            regions=regions,
+            loc=loc,
+            ip=ip,
+        )
+    )
+
   @builtins.property
   def elements(self):
     _ods_variadic_group_length = len(self.operation.operands) - 1 + 1
