diff --git a/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp b/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp
index 85050315..414318eb 100644
--- a/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp
+++ b/enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp
@@ -3940,14 +3940,6 @@ void TypeAnalyzer::visitIntrinsicInst(llvm::IntrinsicInst &I) {
   case Intrinsic::nearbyint:
   case Intrinsic::round:
   case Intrinsic::sqrt:
-#if LLVM_VERSION_MAJOR >= 21
-  case Intrinsic::nvvm_fabs:
-  case Intrinsic::nvvm_fabs_ftz:
-#else
-  case Intrinsic::nvvm_fabs_f:
-  case Intrinsic::nvvm_fabs_d:
-  case Intrinsic::nvvm_fabs_ftz_f:
-#endif
   case Intrinsic::fabs:
     // No direction check as always valid
     updateAnalysis(
