diff --git a/llvm/include/llvm/ADT/STLExtras.h b/llvm/include/llvm/ADT/STLExtras.h
index 3c18aae1b51..b96f2610eeb 100644
--- a/llvm/include/llvm/ADT/STLExtras.h
+++ b/llvm/include/llvm/ADT/STLExtras.h
@@ -1192,10 +1192,11 @@ public:
   const BaseT &getBase() const { return base; }
 
 protected:
+  indexed_accessor_iterator() = default;
   indexed_accessor_iterator(BaseT base, ptrdiff_t index)
       : base(base), index(index) {}
   BaseT base;
-  ptrdiff_t index;
+  ptrdiff_t index = 0;
 };
 
 namespace detail {
@@ -1219,6 +1220,8 @@ public:
   class iterator : public indexed_accessor_iterator<iterator, BaseT, T,
                                                     PointerT, ReferenceT> {
   public:
+    iterator() = default;
+
     // Index into this iterator, invoking a static method on the derived type.
     ReferenceT operator*() const {
       return DerivedT::dereference_iterator(this->getBase(), this->getIndex());
