--- build/python_packages/quantum/mlir_quantum/ir.py	2023-03-07 15:06:48.523539871 -0500
+++ build/python_packages/quantum/mlir_quantum/ir.py.backup	2023-03-07 15:05:13.140767729 -0500
@@ -2,5 +2,5 @@
 #  See https://llvm.org/LICENSE.txt for license information.
 #  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-from ._mlir_libs._mlir.ir import *
-from ._mlir_libs._mlir.ir import _GlobalDebug
+from jax.interpreters.mlir.ir import *
+from jax.interpreters.mlir.ir import _GlobalDebug
